### PR TITLE
Add updateDocuments API which accept a query

### PR DESCRIPTION
**Pull request by zhaih**
### Description

Essentially just wire up `updateDocuments(Query, Iterable<Document>)` API, we could probably also add `updateDocument` for single document update, but I don't see a good reason to add that (except for completion) yet.

### CODE REVIEW

**Comment by uschindler** [comment]
To me this looks fine, although I wonder why we needed a new `Node<T>` implementation, because `deleteDocuments(Query)` was already using a Query to delete. Or are those `Node<T>` instances only used for block indexing?

**Comment by uschindler** [comment]
Sorry I pressed merge on wrong PR, so I reverted this.

Can you reopen this?

**Comment by uschindler** [comment]
Of course you can also revert the revert, if that's applicable.

The change looks fine to me.

**Review by mikemccand — COMMENTED** [review]
Looks great!  I left a couple minor comments.  Thanks @zhaih.

**Comment by zhaih** [comment]
> I wonder why we needed a new Node<T> implementation

@uschindler We previously have a `Node<Query[]>` impl as previous `deleteDocuments` took `(Query...)`. I added that `Node<Query>` just to comply with the `updateDocuments(Term delTerm...)`

File: lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java [hunk 1]
```diff
@@ -3476,7 +3476,12 @@ public int numDeletesToMerge(
                       Document doc = new Document();
                       doc.add(new StringField("id", id, Field.Store.YES));
                       if (mixDeletes && random().nextBoolean()) {
-                        writer.updateDocuments(new Term("id", id), Arrays.asList(doc, doc));
+                        if (random().nextBoolean()) {
+                          writer.updateDocuments(new Term("id", id), Arrays.asList(doc, doc));
+                        } else {
+                          writer.updateDocuments(
```
[the remark this answers is not in the document]
Comment (zhaih) [reply lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java:3482]: I added to it, not sure whether that's the place in your mind?

**Comment by zhaih** [comment]
For some reason I cannot reopen this one, so I opened a new one #12346

**Pull request merged**